### DISCUSSION TITLE
Empty functions or interfaces?

### DISCUSSION THREAD

**Discussion by AbdelrhmanAmin** [reduxjs/redux#4399] [General]
I have noticed a re-occurring pattern in Redux that I don't understand.
What are those empty functions before the actual function? The actual function starts after the red line I drew on the screenshot.

#### `applyMiddleware.ts`

![image](https://user-images.githubusercontent.com/37663043/183394734-6ca6f0b0-ffb5-49f0-b01f-f3f68178140c.png)
---------------------------

#### `compose.ts`

![image](https://user-images.githubusercontent.com/37663043/183395293-07ec5710-f3bd-42a5-9b34-9646756e5b4c.png)

Thanks.

**Comment by markerikson** [comment]
That's TypeScript "method overloading".  It tells TS that there are multiple valid ways to call those functions, with varying arguments and signatures, even though there's only one JS implementation:

- https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads
- https://dmitripavlutin.com/typescript-function-overloading/

**Comment by AbdelrhmanAmin** [comment]
Thank you :)
